src/vsr/grid_scrubber.zig

```diff
@@ -636,3 +636,133 @@ fn ManifestBlockIteratorType(comptime ManifestLog: type) type {
+    // Note that since this parameter is separate from the faults/year rate, increasing
+    // `storage_size` actually reduces the likelihood of data loss.
+    const storage_size = 16 * (1024 * 1024 * 1024 * 1024);
+
+    // The expected (average) number of sector faults per year.
+    // I can't find any good, recent statistics for faults on SSDs.
+    //
+    // Most papers express the fault rate as "UBER" (uncorrectable bit errors per total bits read).
+    // But "Flash Reliability in Production: The Expected and the Unexpected" §5.1 finds that
+    // UBER's underlying assumption ­ that the uncorrectable errors is correlated to the number of
+    // bytes read ­ is false. (That paper only shares "fraction of drives affected by an error",
+    // which is too coarse for this model's purposes.)
+    //
+    // Instead, the parameter is chosen conservatively ­ greater than the "true" number by at least
+    // an order of magnitude.
+    const sector_faults_per_year = 10_000;
+
+    // A block has multiple sectors. If any of a block's sectors are corrupt, then the block is
+    // corrupt.
+    //
+    // Increasing this parameter increases the likelihood of eventual data loss.
+    // (Intuitively, a single bitrot within 1GiB is more likely than a single bitrot within 1KiB.)
+    const block_size = 512 * 1024;
+
+    // The total number of copies of each sector.
+    // The cluster is recoverable if a sector's number of faults is less than `replicas_total`.
+    // Set to 3 rather than 6 since 3 is the quorum_replication.
+    const replicas_total = 3;
+
+    const sector_size = constants.sector_size;
+
+    // Computation:
+
+    const block_sectors = @divExact(block_size, sector_size);
+    const storage_sectors = @divExact(storage_size, sector_size);
+    const storage_blocks = @divExact(storage_size, block_size);
+    const test_duration_days = test_duration_years * 365;
+    const test_duration_cycles = stdx.div_ceil(test_duration_days, cycle_interval_days);
+    const sector_faults_per_cycle =
+        stdx.div_ceil(sector_faults_per_year * cycle_interval_days, 365);
+
+    // P(a specific sector is uncorrupted for an entire cycle)
+    //
+    // If we inject a fault randomly into S sectors, then the probability that the fault misses one
+    // sector in specific is `(S-1)/S`. If we then repeat this N times, then the chance that it
+    // misses that sector every time is `((S-1)/S)^N`.
+    const p_sector_healthy_per_cycle = std.math.pow(
+        f64,
+        @as(f64, @floatFromInt(storage_sectors - 1)) / @as(f64, @floatFromInt(storage_sectors)),
+        @as(f64, @floatFromInt(sector_faults_per_cycle)),
+    );
+
+    // P(a specific block is uncorrupted for an entire cycle)
+    // If any of the block's sectors is corrupted, then the whole block is corrupted.
+    const p_block_healthy_per_cycle =
+        std.math.pow(f64, p_sector_healthy_per_cycle, @as(f64, @floatFromInt(block_sectors)));
```

Comment:
I think this is not pedantically correct. A _different_ way to compute this is to say that a probability that a single fault misses the block is `(S - block_sectors) / S`, and hence the probability that the all faults miss the block is  `((S - block_sectors) / S) ^ sector_faults_per_cycle`, which is a different number, so there must be a difference here!

I think the bug is in the formula in the source code, as you can't multiply probabilities for different sectors, as they are not independent --- if one sector is know to not be faulty then other sectors are more likely to be faulty, assuming we have a constant error rate. 

But it another way, if we expect ten faults per cycle, and our block size is equal to the data size, then we know that the block is correpted, and that the probability of it being healthy is zero. However, probability of every particular sector to be healthy is non-zero, so, if we multiply them all, we'll get something bigger than zero. 

Now, I am not sure if this is actually the problem in the formula, or the problem in how we state the error rates. I guess, the problem is that we say that the faults are uncorrelated, but then use a model where a bag of faults is distributed across the disk, which makes them correlated!